src/passes/InstrumentBranchHints.cpp

```diff
@@ -0,0 +1,334 @@
+// by scanning the inside of the condition. Using that, the new logging will be
+// 123,B,C followed by 123,B,!C. We can therefore find pairs of loggings with
+// the same ID, and consider the predicted and actual values:
+//
+//  [id,0,0], [id,0,0] - nothing changed: good
+//  [id,0,0], [id,0,1] - the actual result changed but not the prediction: bad
+//  [id,0,0], [id,1,0] - prediction changed but not actual result: bad
+//  [id,0,0], [id,1,1] - actual and predicted both changed: good
+//  etc.
+//
+// To make it easy to pair the results, the ID is negative in subsequent
+// instrumentations. That is, we will match an ID of 42 in the first
+// instrumentation with an id of -42 in the second (that avoids us matching two
+// from the first, if e.g. a branch happens twice in a loop). Thus, the first
+// instrumentations adds positive IDs, and the second adds negative, which makes
+// it trivial to differentiate them. (See script/fuzz_opt.py's
+// BranchHintPreservation for more details.)
+//
+
+#include "ir/eh-utils.h"
+#include "ir/local-graph.h"
+#include "ir/names.h"
+#include "ir/properties.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+// The branch id, which increments as we go.
+int branchId = 1;
+
+struct InstrumentBranchHints
+  : public WalkerPass<PostWalker<InstrumentBranchHints>> {
+
+  using Super = WalkerPass<PostWalker<InstrumentBranchHints>>;
+
+  // The module and base names of our import.
+  Name MODULE = "fuzzing-support";
+  Name BASE = "log-branch";
+
+  // The internal name of our import.
+  Name LOG_BRANCH;
+
+  // Whether we are the second pass of instrumentation. If so, we only add
+  // logic to parallel existing hints (for each such hint, we emit one with a
+  // negative ID, so they can be paired, as mentioned above).
+  bool secondInstrumentation = false;
+
+  std::unique_ptr<LocalGraph> localGraph;
+
+  void visitIf(If* curr) { processCondition(curr); }
+
+  void visitBreak(Break* curr) {
+    if (curr->condition) {
+      processCondition(curr);
+    }
+  }
+
+  // Track existing calls to our logging, and their gets, so that we can
+  // identify them and add the second instrumentation properly. This map stores
+  // gets that map to such calls, specifically their actual values (the same
+  // value used in the branch, which we want to instrument). We also map tees.
+  std::unordered_map<LocalGet*, Call*> getsOfPriorInstrumentation;
+  std::unordered_map<LocalSet*, Call*> teesOfPriorInstrumentation;
+
+  void visitCall(Call* curr) {
+    if (curr->target != LOG_BRANCH) {
+      return;
+    }
+    // Our logging has 3 fields: id, expected, actual.
+    assert(curr->operands.size() == 3);
+    if (auto* get = curr->operands[2]->dynCast<LocalGet>()) {
+      getsOfPriorInstrumentation[get] = curr;
+    } else if (auto* tee = curr->operands[2]->dynCast<LocalSet>()) {
+      teesOfPriorInstrumentation[tee] = curr;
+    }
```

Comment:
Is it possible for there to be anything besides a get or tee here? Should we have an assertion if not?